File: internal/tests/csharp/UnitTests.cs
using System;
using System.Collections.Generic;
using Xunit;

namespace B4N1Web.Tests
{
    public class BrowserModeTests
    {
        [Fact] public void LightValue() => Assert.Equal("light", BrowserMode.Light.ToString().ToLower());
        [Fact] public void JsValue() => Assert.Equal("js", BrowserMode.JS.ToString().ToLower());
        [Fact] public void RenderValue() => Assert.Equal("render", BrowserMode.Render.ToString().ToLower());
        [Fact] public void Distinct() {
            var s = new HashSet<BrowserMode> { BrowserMode.Light, BrowserMode.JS, BrowserMode.Render };
            Assert.Equal(3, s.Count);
        }
    }
    public class BrowserOptionsTests
    {
        [Fact] public void Defaults() {
            var o = new BrowserOptions();
            Assert.Equal(BrowserMode.Light, o.Mode); Assert.Equal(30, o.Timeout);
            Assert.Equal("B4N1Web-Agent/1.0", o.UserAgent);
        }
        [Fact] public void SetMode() { var o = new BrowserOptions { Mode = BrowserMode.JS }; Assert.Equal(BrowserMode.JS, o.Mode); }
        [Fact] public void SetTimeout() { var o = new BrowserOptions { Timeout = 60 }; Assert.Equal(60, o.Timeout); }
        [Fact] public void AllTogether() {
            var o = new BrowserOptions { Mode = BrowserMode.Render, Timeout = 120, UserAgent = "T" };
            Assert.Equal(BrowserMode.Render, o.Mode); Assert.Equal(120, o.Timeout);
        }
    }
    public class PageTests
    {
        [Fact] public void Construction() {
            var p = new Page { Url = "https://x.com", Markdown = "# Hi\nContent", Links = new List<string> { "a", "b" } };
            Assert.Equal("https://x.com", p.Url); Assert.Equal(2, p.Links.Count);
        }
        [Fact] public void EmptyDefaults() { var p = new Page(); Assert.Equal(string.Empty, p.Url); Assert.Empty(p.Links); }
        [Fact] public void Unicode() { var p = new Page { Markdown = "日本語\n🚀" }; Assert.Contains("日本語", p.Markdown); }
        [Fact] public void ManyLinks() {
            var l = new List<string>(); for (int i = 0; i < 1000; i++) l.Add($"https://x.com/{i}");
            Assert.Equal(1000, new Page { Links = l }.Links.Count);
        }
    }
    public class GetMainContentTests
    {
        [Fact] public void Empty() => Assert.Equal("", new Page { Markdown = "" }.GetMainContent());
        [Fact] public void SingleLine() => Assert.Equal("Only one", new Page { Markdown = "Only one" }.GetMainContent());
        [Fact] public void SkipsHeader() {
            var p = new Page { Markdown = "# Title\n\n## Sub\n\nActual content" };
            var c = p.GetMainContent(); Assert.Contains("Actual content", c); Assert.DoesNotContain("# Title", c);
        }
        [Fact] public void Unicode() {
            var p = new Page { Markdown = "# Título\n\nContenido 🇪🇸" }; Assert.Contains("Contenido", p.GetMainContent());
        }
    }
    public class FindLinksByTextTests
    {
        [Fact] public void Exact() { var p = new Page { Links = new List<string> { "https://x.com/contact" } }; Assert.Single(p.FindLinksByText("contact")); }
        [Fact] public void NoMatch() => Assert.Empty(new Page { Links = new List<string> { "https://x.com/p" } }.FindLinksByText("missing"));
        [Fact] public void CaseInsensitive() => Assert.Single(new Page { Links = new List<string> { "https://EXAMPLE.COM" } }.FindLinksByText("example"));
        [Fact] public void Multiple() {
            var p = new Page { Links = new List<string> { "https://x.com/contact", "https://x.com/contact-us", "https://y.com" } };
            Assert.Equal(2, p.FindLinksByText("contact").Count);
        }
        [Fact] public void EmptySearch() { var p = new Page { Links = new List<string> { "a", "b" } }; Assert.Equal(2, p.FindLinksByText("").Count); }
        [Fact] public void EmptyList() => Assert.Empty(new Page { Links = new List<string>() }.FindLinksByText("x"));
        [Fact] public void Unicode() => Assert.Single(new Page { Links = new List<string> { "https://x.com/日本語" } }.FindLinksByText("日本語"));
    }
    public class AgentBrowserConstructionTests
    {
        [Fact] public void Defaults() { try { using var b = new AgentBrowser(); Assert.NotNull(b); } catch (BinaryNotFoundException) { } }
        [Fact] public void WithMode() { try { var o = new BrowserOptions { Mode = BrowserMode.Render }; using var b = new AgentBrowser(o); Assert.NotNull(b); } catch (BinaryNotFoundException) { } }
        [Fact] public void WithTimeout() { try { var o = new BrowserOptions { Timeout = 60 }; using var b = new AgentBrowser(o); Assert.NotNull(b); } catch (BinaryNotFoundException) { } }
        [Fact] public void CloseMultiple() { try { var b = new AgentBrowser(); b.Close(); b.Close(); Assert.True(true); } catch (BinaryNotFoundException) { } }
    }
    public class AgentBrowserGotoTests
    {
        private static bool HasBinary() { try { using var b = new AgentBrowser(); return true; } catch { return false; } }
        [Fact] public void Success() {
            if (!HasBinary()) return;
            using var b = new AgentBrowser(); var p = b.Goto("https://example.com");
            Assert.Equal("https://example.com", p.Url); Assert.NotEmpty(p.Markdown);
        }
        [Fact] public void InvalidUrlThrows() {
            if (!HasBinary()) return;
            using var b = new AgentBrowser(); Assert.Throws<Exception>(() => b.Goto("not-a-valid-url"));
        }
        [Fact] public void ReturnsPageInstance() {
            if (!HasBinary()) return;
            using var b = new AgentBrowser(); var p = b.Goto("https://example.com"); Assert.IsType<Page>(p);
        }
    }
    public class ErrorTests
    {
        [Fact] public void BinaryNotFoundExceptionMessage() { var ex = new BinaryNotFoundException(); Assert.NotEmpty(ex.Message); }
        [Fact] public void ThrowAndCatch() {
            try { throw new BinaryNotFoundException(); } catch (BinaryNotFoundException e) { Assert.NotEmpty(e.Message); return; }
            Assert.Fail("Should have thrown");
        }
        [Fact] public void InvalidUrlThrowsException() {
            try { using var b = new AgentBrowser(); b.Goto("not-a-valid-url"); Assert.Fail("Should throw"); }
            catch (BinaryNotFoundException) { } catch (Exception e) { Assert.NotEmpty(e.Message); }
        }
        [Fact] public void ErrorTypesDistinguishable() {
            var bnf = new BinaryNotFoundException(); var ex = new Exception("test"); Assert.NotEqual(bnf.GetType(), ex.GetType());
        }
    }
}

File: internal/tests/java/UnitTests.java
package com.b4n1.web;

import java.util.Arrays;

/**
 * Exhaustive tests for B4n1Web Java SDK.
 * SDK is a thin wrapper around the b4n1web binary.
 */
public class UnitTests {

    private static int passed = 0;
    private static int failed = 0;
    private static int skipped = 0;

    public static void main(String[] args) throws Exception {
        System.out.println("===========================================");
        System.out.println("  B4N1WEB JAVA SDK - TESTS");
        System.out.println("===========================================\n");

        // BrowserMode
        run("BrowserMode values", () -> {
            assertEq(BrowserMode.LIGHT.getValue(), "light", "LIGHT");
            assertEq(BrowserMode.JS.getValue(), "js", "JS");
            assertEq(BrowserMode.RENDER.getValue(), "render", "RENDER");
        });
        run("BrowserMode distinct", () -> {
            assertNeq(BrowserMode.LIGHT, BrowserMode.JS, "LIGHT!=JS");
            assertNeq(BrowserMode.LIGHT, BrowserMode.RENDER, "LIGHT!=RENDER");
            assertNeq(BrowserMode.JS, BrowserMode.RENDER, "JS!=RENDER");
        });
        run("BrowserMode valueOf", () -> {
            assertEq(BrowserMode.valueOf("LIGHT"), BrowserMode.LIGHT, "valueOf");
        });
        run("BrowserMode values()", () -> {
            BrowserMode[] modes = BrowserMode.values();
            assertEq(modes.length, 3, "3 modes");
        });

        // BrowserOptions
        run("BrowserOptions defaults", () -> {
            BrowserOptions opts = new BrowserOptions();
            assertEq(opts.getMode(), BrowserMode.LIGHT, "mode");
            assertEq(opts.getTimeout(), 30, "timeout");
            assertEq(opts.getUserAgent(), "B4N1Web-Agent/1.0", "userAgent");
        });
        run("BrowserOptions setMode", () -> {
            BrowserOptions opts = new BrowserOptions();
            opts.setMode(BrowserMode.JS);
            assertEq(opts.getMode(), BrowserMode.JS, "setMode");
        });
        run("BrowserOptions setTimeout", () -> {
            BrowserOptions opts = new BrowserOptions();
            opts.setTimeout(60);
            assertEq(opts.getTimeout(), 60, "setTimeout");
        });
        run("BrowserOptions setUserAgent", () -> {
            BrowserOptions opts = new BrowserOptions();
            opts.setUserAgent("Custom/1.0");
            assertEq(opts.getUserAgent(), "Custom/1.0", "setUserAgent");
        });

        // Page
        run("Page construction", () -> {
            Page p = new Page("https://example.com", "# Hello\nContent",
                Arrays.asList("https://a.com", "https://b.com"));
            assertEq(p.getUrl(), "https://example.com", "URL");
            assertEq(p.getMarkdown(), "# Hello\nContent", "markdown");
            assertEq(p.getLinks().size(), 2, "links count");
        });
        run("Page empty", () -> {
            Page p = new Page();
            assertTrue(p.getUrl() == null || p.getUrl().isEmpty(), "empty URL");
            assertTrue(p.getMarkdown() == null || p.getMarkdown().isEmpty(), "empty markdown");
        });
        run("Page unicode", () -> {
            Page p = new Page("https://example.com/日本語", "日本語テスト\n🚀",
                Arrays.asList("https://例.jp"));
            assertTrue(p.getMarkdown().contains("日本語"), "unicode");
        });
        run("Page setScreenshot", () -> {
            Page p = new Page("u", "Content", Arrays.asList());
            p.setScreenshot("base64");
            assertEq(p.getScreenshot(), "base64", "screenshot");
        });

        // GetMainContent
        run("GetMainContent empty", () -> {
            Page p = new Page("u", "", Arrays.asList());
            assertEq(p.getMainContent(), "", "empty");
        });
        run("GetMainContent single", () -> {
            Page p = new Page("u", "Only one line", Arrays.asList());
            assertEq(p.getMainContent(), "Only one line", "single");
        });
        run("GetMainContent skips header", () -> {
            Page p = new Page("u", "# Title\n\n## Sub\n\nActual content", Arrays.asList());
            String got = p.getMainContent();
            assertTrue(got.contains("Actual content"), "has content");
            assertTrue(!got.contains("# Title"), "skips header");
        });
        run("GetMainContent many lines", () -> {
            StringBuilder sb = new StringBuilder("# Header\n\n## Sub\n");
            for (int i = 0; i < 100; i++) sb.append("Line ").append(i).append("\n");
            Page p = new Page("u", sb.toString(), Arrays.asList());
            assertTrue(p.getMainContent().length() > 50, "many lines");
        });

        // FindLinksByText
        run("FindLinks exact", () -> {
            Page p = new Page("u", "", Arrays.asList("https://x.com/contact"));
            assertEq(p.findLinksByText("contact").size(), 1, "exact");
        });
        run("FindLinks partial", () -> {
            Page p = new Page("u", "", Arrays.asList("https://x.com/about-us"));
            assertEq(p.findLinksByText("about").size(), 1, "partial");
        });
        run("FindLinks no match", () -> {
            Page p = new Page("u", "", Arrays.asList("https://x.com/page"));
            assertEq(p.findLinksByText("missing").size(), 0, "no match");
        });
        run("FindLinks case insensitive", () -> {
            Page p = new Page("u", "", Arrays.asList("https://EXAMPLE.COM"));
            assertEq(p.findLinksByText("example").size(), 1, "case");
        });
        run("FindLinks multiple", () -> {
            Page p = new Page("u", "", Arrays.asList(
                "https://x.com/contact", "https://x.com/contact-us", "https://y.com"));
            assertEq(p.findLinksByText("contact").size(), 2, "multiple");
        });
        run("FindLinks empty search", () -> {
            Page p = new Page("u", "", Arrays.asList("https://a.com", "https://b.com"));
            assertEq(p.findLinksByText("").size(), 2, "empty search");
        });
        run("FindLinks empty list", () -> {
            Page p = new Page("u", "", Arrays.asList());
            assertEq(p.findLinksByText("x").size(), 0, "empty list");
        });
        run("FindLinks unicode", () -> {
            Page p = new Page("u", "", Arrays.asList("https://x.com/日本語"));
            assertEq(p.findLinksByText("日本語").size(), 1, "unicode");
        });

        // AgentBrowser
        run("AgentBrowser defaults", () -> {
            try {
                AgentBrowser b = new AgentBrowser();
                pass("construct");
                b.close();
            } catch (BinaryNotFoundException e) {
                skip("binary not found");
            }
        });
        run("AgentBrowser with mode", () -> {
            try {
                BrowserOptions opts = new BrowserOptions();
                opts.setMode(BrowserMode.JS);
                AgentBrowser b = new AgentBrowser(opts);
                pass("construct with mode");
                b.close();
            } catch (BinaryNotFoundException e) {
                skip("binary not found");
            }
        });
        run("AgentBrowser with timeout", () -> {
            try {
                BrowserOptions opts = new BrowserOptions();
                opts.setTimeout(60);
                AgentBrowser b = new AgentBrowser(opts);
                pass("construct with timeout");
                b.close();
            } catch (BinaryNotFoundException e) {
                skip("binary not found");
            }
        });
        run("AgentBrowser with userAgent", () -> {
            try {
                BrowserOptions opts = new BrowserOptions();
                opts.setUserAgent("Custom/1.0");
                AgentBrowser b = new AgentBrowser(opts);
                pass("construct with userAgent");
                b.close();
            } catch (BinaryNotFoundException e) {
                skip("binary not found");
            }
        });
        run("AgentBrowser close multiple", () -> {
            try {
                AgentBrowser b = new AgentBrowser();
                b.close(); b.close(); b.close();
                pass("multiple closes");
            } catch (BinaryNotFoundException e) {
                skip("binary not found");
            }
        });

        // Goto with real binary
        run("Goto success", () -> {
            try {
                AgentBrowser b = new AgentBrowser();
                Page p = b.goto_("https://example.com");
                assertEq(p.getUrl(), "https://example.com", "URL");
                assertTrue(p.getMarkdown().length() > 0, "markdown");
                assertTrue(p.getLinks() != null, "links");
                b.close();
            } catch (BinaryNotFoundException e) {
                skip("binary not found");
            } catch (Exception e) {
                fail("goto: " + e.getMessage());
            }
        });
        run("Goto JS mode", () -> {
            try {
                BrowserOptions opts = new BrowserOptions();
                opts.setMode(BrowserMode.JS);
                AgentBrowser b = new AgentBrowser(opts);
                Page p = b.goto_("https://example.com");
                assertEq(p.getUrl(), "https://example.com", "URL");
                b.close();
            } catch (BinaryNotFoundException e) {
                skip("binary not found");
            } catch (Exception e) {
                fail("goto JS: " + e.getMessage());
            }
        });
        run("Goto invalid URL", () -> {
            try {
                AgentBrowser b = new AgentBrowser();
                try {
                    b.goto_("not-a-valid-url");
                    fail("invalid URL should throw");
                } catch (NavigationException e) {
                    pass("invalid URL throws");
                }
                b.close();
            } catch (BinaryNotFoundException e) {
                skip("binary not found");
            }
        });
        run("Goto non-existent domain", () -> {
            try {
                AgentBrowser b = new AgentBrowser();
                try {
                    b.goto_("https://this-domain-definitely-does-not-exist-abc123xyz.com");
                    fail("non-existent should throw");
                } catch (NavigationException e) {
                    pass("non-existent throws");
                }
                b.close();
            } catch (BinaryNotFoundException e) {
                skip("binary not found");
            }
        });

        // Version
        run("GetVersion", () -> {
            String ver = AgentBrowser.getVersion();
            if (ver == null || ver.isEmpty() || ver.equals("unknown")) {
                skip("binary not in system path");
            } else {
                pass("version: " + ver);
            }
        });

        // Errors
        run("BinaryNotFoundException", () -> {
            try {
                throw new BinaryNotFoundException("test message");
            } catch (BinaryNotFoundException e) {
                assertTrue(e.getMessage().contains("test message"), "message");
                pass("BinaryNotFoundException");
            }
        });
        run("NavigationException", () -> {
            try {
                throw new NavigationException("https://x.com", "nav error");
            } catch (NavigationException e) {
                assertEq(e.getUrl(), "https://x.com", "URL");
                pass("NavigationException");
            }
        });

        // Results
        System.out.println("\n===========================================");
        System.out.println("  RESULTS: " + passed + " passed, " + failed + " failed, " + skipped + " skipped");
        System.out.println("===========================================");
        if (failed > 0) System.exit(1);
    }

    // Test helpers
    private static void run(String name, Runnable test) {
        try { test.run(); }
        catch (AssertionError e) { System.err.println("  ❌ " + name + ": " + e.getMessage()); failed++; }
        catch (Exception e) { System.err.println("  ❌ " + name + ": " + e.getClass().getSimpleName() + ": " + e.getMessage()); failed++; }
    }
    private static void pass(String m) { passed++; }
    private static void fail(String m) { System.err.println("  ❌ " + m); failed++; }
    private static void skip(String m) { System.out.println("  ⏭️  " + m); skipped++; }
    private static void assertEq(Object a, Object b, String n) {
        if (!a.equals(b)) throw new AssertionError(n + ": expected " + b + ", got " + a);
        passed++;
    }
    private static void assertNeq(Object a, Object b, String n) {
        if (a.equals(b)) throw new AssertionError(n + ": should differ");
        passed++;
    }
    private static void assertTrue(boolean c, String n) {
        if (!c) throw new AssertionError(n + ": expected true");
        passed++;
    }
}
